Update rust bitcoin

danielabrozzoni opened the pull request:
### Description

Updates to rust-bitcoin 0.30.0 and miniscript 0.10.0

Not covered in this PR:
- https://github.com/bitcoindevkit/bdk/issues/1036. Although the latter is deprecated, I think it's better if I update it in a separate PR, as this one is pretty big already. 
- https://github.com/bitcoindevkit/bdk/issues/1037
- https://github.com/bitcoindevkit/bdk/issues/1038

Heads up, I'm explicitly denying multipath descriptors until we have better tests for them. See the commit message of 23fba7aee9b55bf59db73c296b9fb025a256502e

### Changelog notice

- Update to `rust-bitcoin` 0.30.0 and `miniscript` 10.0.0

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

danielabrozzoni committed:
Update bdk_chain to bitcoin 0.30.0

danielabrozzoni committed:
Update crates/esplora to rust-bitcoin 0.30.0

danielabrozzoni committed:
Update crates/electrum to bitcoin 0.30.0

danielabrozzoni committed:
Update wallet_esplora and wallet_esplora_async to...
...rust-bitcoin 0.30.0

danielabrozzoni committed:
Update tmp_plan to rust-bitcoin 0.30.0

danielabrozzoni committed:
Update coin_select to rust-bitcoin 0.30.0

danielabrozzoni committed:
Update example_cli to rust-bitcoin 0.30.0

danielabrozzoni committed:
Update example_electrum to rust-bitcoin 0.30.0

danielabrozzoni committed:
Update wallet_electrum to rust-bitcoin 0.30.0

evanlinjin commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -38,20 +38,22 @@
+//!         fee_rate: bdk::FeeRate,
 //!         target_amount: u64,
-//!         drain_script: &Script,
+//!         drain_script: &ScriptBuf,
```
Instead of `&ScriptBuf`, consider having `&Script`.

evanlinjin commented on crates/chain/src/chain_data.rs:
```diff
@@ -104,7 +104,7 @@ impl Default for BlockId {
         Self {
             height: Default::default(),
-            hash: BlockHash::from_inner([0u8; 32]),
+            hash: BlockHash::from_slice(&[0u8; 32]).expect("len is 32"),
```
Consider using `BlockHash::all_zeros()`.

apoelstra replied:
Also, `from_inner` was renamed to `from_byte_array`, so for other fixed arrays you still don't need to `.expect`.

apoelstra commented on crates/chain/tests/test_spk_txout_index.rs:
```diff
@@ -1,18 +1,21 @@
     let tx1 = Transaction {
         version: 0x02,
-        lock_time: PackedLockTime(0),
+        lock_time: LockTime::Blocks(Height::ZERO),
```
In e83ea8e1204833049e8d1dd5c6b2be4c981e463d:

There is a `Locktime::ZERO` constant you can use directly.

Also, as a style thing, I find it clearer to import `absolute` and `relative` modules and then to use `absolute::LockTime` everywhere so it's clear what kind of locktime is under consideration.

apoelstra commented on crates/bdk/src/descriptor/dsl.rs:
```diff
@@ -702,7 +703,7 @@ macro_rules! fragment {
     });
     ( after ( $value:expr ) ) => ({
-        $crate::impl_leaf_opcode_value!(After, $crate::bitcoin::PackedLockTime($value)) // TODO!! https://github.com/rust-bitcoin/rust-bitcoin/issues/1302
+        $crate::impl_leaf_opcode_value!(After, $crate::bitcoin::absolute::LockTime::from_consensus($value).into()) // TODO!! https://github.com/rust-bitcoin/rust-bitcoin/issues/1302
```
In d12242ec746ffc51da64672d7c7f7e014258d190:

You can do `$crate::miniscript::AbsLockTime::from_consensus($value)` which is much shorter than this and avoids the `.into()`. Can probably also drop the TODO at that point unless you want to file an issue on rust-miniscript suggesting that `AbsLockTIme` implement `From<u32>`.

apoelstra commented on crates/bdk/src/descriptor/mod.rs:
```diff
@@ -494,7 +512,7 @@ impl DescriptorMeta for ExtendedDescriptor {
         });
 
-        path_found.map(|path| self.at_derivation_index(path))
+        path_found.map(|path| self.at_derivation_index(path).expect("TODO"))
```
In d12242ec746ffc51da64672d7c7f7e014258d190:

This TODO can just be `unwrap` with a comment explaining that `path_found` came out of a child index and therefore is in range.

apoelstra commented on crates/bdk/src/descriptor/mod.rs:
```diff
@@ -545,7 +563,7 @@ impl DescriptorMeta for ExtendedDescriptor {
         }
 
-        let descriptor = self.at_derivation_index(0);
+        let descriptor = self.at_derivation_index(0).expect("TODO");
```
In d12242ec746ffc51da64672d7c7f7e014258d190:

You can just unwrap. This function only fails if the index is out of range, and 0 is not.

apoelstra replied:
Ditto at several other places.

apoelstra commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1403,7 +1404,7 @@ impl<D> Wallet<D> {
             .index_of_spk(&txout.script_pubkey)?;
         let descriptor = self.get_descriptor_for_keychain(keychain);
-        Some(descriptor.at_derivation_index(child))
+        Some(descriptor.at_derivation_index(child).expect("TODO"))
```
In d12242ec746ffc51da64672d7c7f7e014258d190:

These TODOs could be `.ok()?` which wil ljust return None if the child index is somehow out of range.

apoelstra commented on example-crates/example_electrum/src/main.rs:
```diff
@@ -84,6 +84,7 @@ fn main() -> anyhow::Result<()> {
         Network::Testnet => "ssl://electrum.blockstream.info:60002",
         Network::Regtest => "tcp://localhost:60401",
         Network::Signet => "tcp://signet-electrumx.wakiyamap.dev:50001",
+        _ => todo!(),
```
In a219219691bb1c6d47e8f8a9df3100f92eee0baf:

`todo` is not a bad choice for this but I'd panic with "Unknown network" or something so it's clear this isn't a code todo.

apoelstra commented:
Done reviewing 81ef698ed9c1fdc6dee1bc2757fa7c9209273343.

Overall looks great. I apologize for what an effort this was.

notmandatory commented:
This should fix #916

tcharding commented on crates/bdk/Cargo.toml:
```diff
@@ -15,8 +15,9 @@ rust-version = "1.57"
-bitcoin = { version = "0.29", features = ["serde", "base64", "rand"], default-features = false }
+miniscript = { version = "10", features = ["serde"], default-features = false }
+bitcoin = { version = "0.30", features = ["serde", "base64", "rand"], default-features = false }
+secp256k1 = { version = "0.27", features = [ "rand-std" ] }
```
```suggestion
bitcoin = { version = "0.30", features = ["serde", "base64", "rand-std"], default-features = false }
```
I guess the secp dependency was just to enable "rand-std"? `rust-bitcoin` has "rand-std" that can do that for you.

tcharding commented on crates/bdk/src/descriptor/mod.rs:
```diff
@@ -600,7 +617,7 @@ mod test {
         )
         .unwrap();
-        let psbt: psbt::PartiallySignedTransaction = deserialize(
+        let psbt = psbt::PartiallySignedTransaction::deserialize(
```
```suggestion
        let psbt = psbt::Psbt::deserialize(
```

`bitcoin::psbt::Psbt` is available now. Next release hopefully `bitcoin::Psbt` will be (https://github.com/rust-bitcoin/rust-bitcoin/pull/1938)

tcharding commented on crates/bdk/src/descriptor/policy.rs:
```diff
@@ -68,7 +68,7 @@ use crate::wallet::utils::{After, Older, SecpCtx};
 use super::error::Error;
 use super::XKeyUtils;
-use bitcoin::util::psbt::{Input as PsbtInput, PartiallySignedTransaction as Psbt};
+use bitcoin::psbt::{Input as PsbtInput, PartiallySignedTransaction as Psbt};
```
Stylistic choice but you could use `psbt::Input` and `Psbt`. Along with:

```
use bitcoin::psbt::{self, Psbt};
```

tcharding commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -38,20 +38,22 @@
 //!     ) -> Result<CoinSelectionResult, bdk::Error> {
 //!         let mut selected_amount = 0;
-//!         let mut additional_weight = 0;
+//!         let mut additional_weight = Weight::from_wu(0);
```
```suggestion
//!         let mut additional_weight = Weight::ZERO;
```

tcharding commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -188,7 +193,7 @@ pub trait CoinSelectionAlgorithm: core::fmt::Debug {
         fee_rate: FeeRate,
         target_amount: u64,
-        drain_script: &Script,
+        drain_script: &ScriptBuf,
```
@evanlinjin already mentioned it but just FYI the `ScriptBuf` is to `Script` as `PathBuf` is to `Path` and `String` is to `&str` - so typically one would always using either `&Script` or `ScriptBuf` depending on whether you want to own the script data or not.

tcharding commented on crates/bdk/src/wallet/signer.rs:
```diff
@@ -499,12 +550,12 @@ fn sign_psbt_ecdsa(
     allow_grinding: bool,
 ) {
-    let msg = &Message::from_slice(&hash.into_inner()[..]).unwrap();
+    let msg = &Message::from_slice(&hash[..]).unwrap();
```
I think you can just do `Message::from(hash)`.

tcharding commented on crates/bdk/src/wallet/signer.rs:
```diff
@@ -523,12 +574,10 @@ fn sign_psbt_schnorr(
     psbt_input: &mut psbt::Input,
-    hash: taproot::TapSighashHash,
-    hash_ty: SchnorrSighashType,
+    hash: sighash::TapSighash,
```
Stylistic choice but I'd do 
```suggestion
    hash: TapSighash,
```

tcharding commented on crates/bdk/src/wallet/signer.rs:
```diff
@@ -849,19 +903,26 @@ impl ComputeSighash for Legacy {
 }
 
-fn p2wpkh_script_code(script: &Script) -> Script {
+fn p2wpkh_script_code(script: &ScriptBuf) -> ScriptBuf {
```
We have `ScriptBuf::p2wpkh_script_code` now (I actually took the code from bdk when adding it :)

tcharding commented on crates/esplora/Cargo.toml:
```diff
@@ -13,7 +13,7 @@ readme = "README.md"
 [dependencies]
 bdk_chain = { path = "../chain", version = "0.4.0", default-features = false, features = ["serde", "miniscript"] }
-esplora-client = { version = "0.5", default-features = false }
+esplora-client = { git = "https://github.com/tcharding/rust-esplora-client.git", branch = "06-10-upgrade-bitcoin", default-features = false }
```
I have my branches cofigured to auto-delete on merge so I changed this to
```
esplora-client = { git = "https://github.com/bitcoindevkit/rust-esplora-client.git", branch = "master", default-features = false }
```

tcharding commented on crates/bdk/src/wallet/signer.rs:
```diff
@@ -849,19 +903,26 @@ impl ComputeSighash for Legacy {
         .into_script()
 }
 
 impl ComputeSighash for Segwitv0 {
```
I'm not exactly sure bit it looks like the implementations of this `ComputeSighash` trait contain loads of stuff that should be done by `rust-bitcoin::psbt` (complex enough to require a separate PR though). The taproot stuff is not done yet but for legacy and segwitv0 you should be able to use the new signing stuff in `psbt` (see `Psbt::sign` and `Psbt::sighash_ecdsa`). If you do play around with it and find it lacking please raise an issue.

danielabrozzoni replied:
Fair, thanks! This PR is big enough already, but I opened https://github.com/bitcoindevkit/bdk/issues/1038 so that I can't forget :)

tcharding replied:
Awesome, thanks.

tcharding commented on crates/bdk/src/wallet/signer.rs:
```diff
@@ -912,7 +973,7 @@ impl ComputeSighash for Segwitv0 {
                     .redeem_script
                     .as_ref()
-                    .map(Script::is_v0_p2wpkh)
+                    .map(|s| s.as_script().is_v0_p2wpkh())
```
No need to call `as_script` explicitly, `ScriptBuf` derefs to `Script` so all the methods on `Script` are available on `ScriptBuf`.

tcharding commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -27,7 +31,7 @@
 //!     .add_recipient(to_address.script_pubkey(), 50_000)
 //!     // With a custom fee rate of 5.0 satoshi/vbyte
-//!     .fee_rate(FeeRate::from_sat_per_vb(5.0))
+//!     .fee_rate(bdk::FeeRate::from_sat_per_vb(5.0))
```
I didn't look super closely but you should be able to get rid of your `FeeRate` now and use `bitcoin::FeeRate`.

danielabrozzoni replied:
I opened an issue for this one: https://github.com/bitcoindevkit/bdk/issues/1037 as I feel it needs more discussion: at the moment `bdk::FeeRate` has a couple methods that `bitcoin::FeeRate` doesn't have, and although switching is possible, I think we should discuss whether it's better to switch immediately or to first submit a patch to you guys :)

tcharding commented on crates/bdk/tests/wallet.rs:
```diff
@@ -725,7 +729,7 @@ fn test_create_tx_set_redeem_witness_script_p2wsh_p2sh() {
     let (psbt, _) = builder.finish().unwrap();
 
-    let script = Script::from(
+    let script = ScriptBuf::from(
```
Can just call `SciptBuf::from_hex` directly here.

tcharding commented on crates/bdk/tests/wallet.rs:
```diff
@@ -809,7 +813,10 @@ fn test_create_tx_add_utxo() {
+    let addr = Address::from_str("2N1Ffz3WaNzbeLFBb51xyFMHYSEUXcbiSoX")
+        .unwrap()
+        .require_network(Network::Testnet)
+        .unwrap();
```
In test code, if you want, you can just call `assumed_checked` instead of `require_network`.

tcharding commented on crates/chain/src/spk_iter.rs:
```diff
@@ -151,12 +150,25 @@ mod test {
-        let external_spk_20 = external_desc.at_derivation_index(20).script_pubkey();
-        let external_spk_21 = external_desc.at_derivation_index(21).script_pubkey();
+        let external_spk_0 = external_desc
+            .at_derivation_index(0)
```
Just fyi if you want to keep these all on one line you can use `#[rustfmt::skip]` on the function.

tcharding commented on nursery/tmp_plan/src/lib.rs:
```diff
@@ -16,15 +16,17 @@
-    blockdata::{locktime::LockTime, transaction::Sequence},
+    address::WitnessVersion,
+    bip32::{DerivationPath, Fingerprint, KeySource},
+    // TODO: is locktime absolute?
```
Did you want this in there still?

tcharding commented on nursery/tmp_plan/src/lib.rs:
```diff
@@ -201,7 +203,7 @@ where
                 } => PlanState::Complete {
                     final_script_sig: None,
-                    final_script_witness: Some(Witness::from_vec(witness)),
+                    final_script_witness: Some(Witness::from_slice(&witness)),
```
```suggestion
                    final_script_witness: Some(Witness::from(witness)),
```

tcharding reviewed (commented):
Wow you guys really use `rust-bitcoin` a lot. I hope we can stabalize the API soon so you don't have to go through all this work too many times. Thanks for being patient with our wild API changes.

danielabrozzoni committed:
Update bdk to bitcoin 0.30.0

danielabrozzoni committed:
Explicitly deny multipath keys

Although there is *some* code to handle multipath keys inside bdk,
it's all untested, and from a few quick tests it
seems that it's pretty easy to find buggy edge cases.
Better to deny multipath descs for now, and revisit the
decision once we work on supporting multidescriptor wallets.

danielabrozzoni commented:
Thanks for the reviews, they've been super helpful! 

I just updated the code and rebased on top of master. CI should be fine now.

apoelstra reviewed (approved):
utACK 6259639a680161ac78caabee5a9a6a84e866c1a7

tcharding commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -152,11 +152,11 @@ impl<I: Clone + Ord> SpkTxOutIndex<I> {
         use core::ops::Bound::*;
         let min_op = OutPoint {
-            txid: Txid::from_inner([0x00; 32]),
+            txid: Txid::from_slice(&[0x00; 32]).expect("len is 32"),
```
nit: `Txid::all_zeros()`

tcharding commented on crates/bdk/src/wallet/signer.rs:
```diff
@@ -499,12 +548,12 @@ fn sign_psbt_ecdsa(
     allow_grinding: bool,
 ) {
-    let msg = &Message::from_slice(&hash.into_inner()[..]).unwrap();
+    let msg = &Message::from_slice(&hash[..]).unwrap();
```
Missed this one, you did the one below :)
```rust
let msg = &Message::from(hash);
```

tcharding reviewed (approved):
utACK 23fba7aee9b55bf59db73c296b9fb025a256502e

tcharding commented:
Clean! Nice one.

evanlinjin commented on crates/bdk/src/descriptor/dsl.rs:
```diff
@@ -824,16 +824,16 @@ mod test {
             } else {
-                desc.at_derivation_index(index)
+                desc.at_derivation_index(index).unwrap()
             };
```
**Nit:** I think this can be simplified to this.

```rust
let child_desc = desc
    .at_derivation_index(i as u32)
    .expect("must not be hardened");
```

For non-wildcard or single keys, `at_derivation_index` basically just wraps the original key in `DefiniteDescriptorKey` (and `index` is not used anyway).

evanlinjin commented on crates/bdk/src/descriptor/error.rs:
```diff
@@ -64,6 +66,7 @@ impl fmt::Display for Error {
                 f,
                 "The descriptor contains hardened derivation steps on public extended keys"
             ),
+            Self::MultiPath => write!(f, "The descriptor contains multipath keys"),
```
```suggestion
            Self::MultiPath => write!(f, "The descriptor contains multipath keys, which are not supported yet"),
```

I think it makes sense to make it explicit that mutipath keys aren't inherently problematic, it's just we don't support it yet.

evanlinjin commented on crates/bdk/src/descriptor/template.rs:
```diff
@@ -565,30 +565,30 @@ mod test {
 
         if let ExtendedDescriptor::Pkh(pkh) = xdesc.0 {
-            let path: Vec<ChildNumber> = pkh.into_inner().full_derivation_path().into();
+            let path: Vec<ChildNumber> = pkh.into_inner().full_derivation_path().unwrap().into();
```
```suggestion
            let path: Vec<ChildNumber> = pkh.into_inner().full_derivation_path().expect("must not be multipath key").into();
```

danielabrozzoni replied:
I'm all in for using expect more often, but I don't think I'll do it in this specific case, as it's a test and it's obvious from the descriptor (just a couple of lines above) that it's not a multipath. But yeah, I'm replacing unwrap with expect in the rest of the code :)

evanlinjin commented on nursery/coin_select/src/coin_selector.rs:
```diff
@@ -94,18 +94,14 @@ impl CoinSelectorOpt {
-            tx.weight() - base_weight
-        };
+        let base_weight = tx.weight().to_wu();
+        let drain_weight = drain_output.weight();
```
My only concern with this, is this weight calculation does not take into account varint changes for output count.

danielabrozzoni replied:
Ahh right, I missed that, thanks!

evanlinjin commented on crates/electrum/Cargo.toml:
```diff
@@ -13,4 +13,6 @@ readme = "README.md"
-electrum-client = { version = "0.12" }
+electrum-client = { version = "0.17" }
+# rustls 0.21.2 increased their MSRV
+rustls = { version = "=0.21.1", optional = true, features = ["dangerous_configuration"] }
```
I think it will make sense to use #1046 here.

danielabrozzoni replied:
Ugh, of course 🤦🏻‍♀️

evanlinjin commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -152,11 +152,11 @@ impl<I: Clone + Ord> SpkTxOutIndex<I> {
         };
         let max_op = OutPoint {
-            txid: Txid::from_inner([0xff; 32]),
+            txid: Txid::from_slice(&[0xff; 32]).expect("len is 32"),
```
Nit:

```suggestion
            txid: Txid::from_byte_array([0xff; Txid::LEN]),
```

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1403,13 +1407,14 @@ impl<D> Wallet<D> {
         self.list_unspent()
             .map(|utxo| {
                 let keychain = utxo.keychain;
+                #[allow(deprecated)]
```
This is a good call. Saves time.

evanlinjin reviewed (commented):
Thank you for this work!

Can I suggest to use `.expect()` instead of `.unwrap()` in some places?

I understand that it becomes tedious to do it everywhere, especially if there are multiple calls to `.at_derivation_index()` (in which case I would suggest to leave a comment explaining why we don't expect a hardened wildcard).

An explanation as to why we do not expect an error goes a long way.

evanlinjin reviewed (changes requested):
I have looked through everything and tested the examples and made sure they all still work :smile:

In my opinion, these are mandatory changes:

* Do not pin dependencies: https://github.com/bitcoindevkit/bdk/pull/1023#discussion_r1281591844
* Better description for multipath error: https://github.com/bitcoindevkit/bdk/pull/1023#discussion_r1281446160
* Incorrect weight calculation: https://github.com/bitcoindevkit/bdk/pull/1023#discussion_r1281584907

After which, I will give an ACK.

danielabrozzoni committed:
ci: Pin rustls to keep the MSRV

danielabrozzoni commented:
Rebased and fixed review comments.

evanlinjin reviewed (approved):
ACK 1da3b304bb0aa29e3bb013e0a583c34ec390b32d

Looked through everything and made sure the examples still worked as expected (which they did).
